File: frontend/src/services/music-media.ts
import { http } from '@/services/http';
import type { EntityType, MetadataDetail } from '@/types/metadata';
import type { ApiMusicResolveDetailResponse, MusicMediaInfo, MusicMediaInput } from '@/types/music-media';

function resolveMusicBrainzRefKey(entityType: EntityType): string | null {
  if (entityType === 'artist') {
    return 'musicbrainz_artist_id';
  }
  if (entityType === 'album') {
    return 'musicbrainz_release_group_id';
  }
  if (entityType === 'track') {
    return 'musicbrainz_recording_id';
  }
  return null;
}

export function buildMusicMediaInputFromMetadataDetail(
  detail: MetadataDetail,
  sourceKind: string,
  sourceContext: Record<string, unknown> = {},
): MusicMediaInput {
  const externalRefs: Record<string, string> = { ...(detail.external_ids ?? {}) };
  const musicBrainzRefKey = detail.provider === 'musicbrainz' ? resolveMusicBrainzRefKey(detail.entity_type) : null;
  if (musicBrainzRefKey && detail.id) {
    externalRefs[musicBrainzRefKey] = detail.id;
  }

  return {
    entity_hint: detail.entity_type,
    source_kind: sourceKind,
    title: detail.track_title || detail.title,
    subtitle: detail.note,
    artist_names: detail.artist_name ? [detail.artist_name] : [],
    album_title: detail.album_title ?? null,
    album_artist_names: [],
    release_date: null,
    year: detail.year ?? null,
    track_number: null,
    disc_number: null,
    external_refs: externalRefs,
    source_context: sourceContext,
    raw_context: {},
  };
}

export function buildMusicMediaInputFromMusicMediaInfo(
  media: MusicMediaInfo,
  sourceKind: string,
  sourceContext: Record<string, unknown> = {},
): MusicMediaInput {
  const externalRefs: Record<string, string> = { ...(media.external_refs ?? {}) };
  const musicBrainzRefKey = media.provider === 'musicbrainz' ? resolveMusicBrainzRefKey(media.entity_type) : null;
  if (musicBrainzRefKey && media.provider_id) {
    externalRefs[musicBrainzRefKey] = media.provider_id;
  }

  return {
    entity_hint: media.entity_type,
    source_kind: sourceKind,
    title: media.title ?? null,
    subtitle: media.match_strategy ?? null,
    artist_names: [...media.artist_names],
    album_title: media.album_title ?? null,
    album_artist_names: [...media.album_artist_names],
    release_date: media.release_date ?? null,
    year: media.year ?? null,
    track_number: media.track_number ?? null,
    disc_number: media.disc_number ?? null,
    external_refs: externalRefs,
    source_context: sourceContext,
    raw_context: {},
  };
}

export async function resolveMusicMediaDetail(input: MusicMediaInput): Promise<ApiMusicResolveDetailResponse> {
  const { data } = await http.post<ApiMusicResolveDetailResponse>('/media/resolve/detail', {
    input,
  });
  return data;
}

File: plugin_runtime/plugins/musicpilot/schemas/music_media.py
"""Core domain models for the unified music media chain."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from .metadata import MetadataDetail
from .mvp import EntityType


class MusicMediaSourceKind(str, Enum):
    DISCOVERY = "discovery"
    DETAIL = "detail"
    SEARCH = "search"
    SUBSCRIPTION = "subscription"
    SUBSCRIPTION_DETAIL = "subscription_detail"
    SUBSCRIPTION_RESOLUTION = "subscription_resolution"
    ACQUISITION = "acquisition"
    LIBRARY = "library"
    MANUAL = "manual"


class MusicRecognitionState(str, Enum):
    DIRECT = "direct"
    READY = "ready"
    INSUFFICIENT = "insufficient"
    FAILED = "failed"


class MusicMediaMatchStrategy(str, Enum):
    STRONG_REF = "strong_ref"
    METADATA_SEARCH = "metadata_search"


class MusicMediaInput(BaseModel):
    """Raw music clues collected from an upstream scenario."""

    entity_hint: EntityType | None = None
    source_kind: MusicMediaSourceKind
    title: str | None = None
    subtitle: str | None = None
    artist_names: list[str] = Field(default_factory=list)
    album_title: str | None = None
    album_artist_names: list[str] = Field(default_factory=list)
    release_date: str | None = None
    year: int | None = None
    track_number: int | None = None
    disc_number: int | None = None
    external_refs: dict[str, str] = Field(default_factory=dict)
    source_context: dict[str, Any] = Field(default_factory=dict)
    raw_context: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="forbid")


class MusicMetaBase(BaseModel):
    """Normalized music metadata before final entity recognition."""

    entity_type: EntityType
    canonical_title: str | None = None
    canonical_artist_names: list[str] = Field(default_factory=list)
    canonical_album_title: str | None = None
    canonical_album_artist_names: list[str] = Field(default_factory=list)
    canonical_release_date: str | None = None
    canonical_year: int | None = None
    track_number: int | None = None
    disc_number: int | None = None
    alias_titles: list[str] = Field(default_factory=list)
    alias_artist_names: list[str] = Field(default_factory=list)
    alias_album_titles: list[str] = Field(default_factory=list)
    featuring_artist_names: list[str] = Field(default_factory=list)
    external_refs: dict[str, str] = Field(default_factory=dict)
    source_refs: dict[str, str] = Field(default_factory=dict)
    evidence: list[dict[str, Any]] = Field(default_factory=list)
    normalization_notes: list[str] = Field(default_factory=list)
    confidence_hint: float | None = None

    model_config = ConfigDict(extra="forbid")


class MusicMediaInfo(BaseModel):
    """Resolved formal music media object used by downstream flows."""

    entity_type: EntityType
    provider: str
    provider_id: str
    title: str | None = None
    artist_names: list[str] = Field(default_factory=list)
    album_title: str | None = None
    album_artist_names: list[str] = Field(default_factory=list)
    release_date: str | None = None
    year: int | None = None
    track_number: int | None = None
    disc_number: int | None = None
    related_artist_ids: list[str] = Field(default_factory=list)
    related_album_id: str | None = None
    related_track_ids: list[str] = Field(default_factory=list)
    external_refs: dict[str, str] = Field(default_factory=dict)
    match_confidence: float | None = None
    match_strategy: MusicMediaMatchStrategy | None = None
    match_evidence: list[dict[str, Any]] = Field(default_factory=list)
    diagnostics: list[str] = Field(default_factory=list)
    cover_url: str | None = None
    disambiguation: str | None = None
    release_context: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="forbid")


class MusicRecognitionAssessment(BaseModel):
    """Recognition readiness summary emitted by the unified chain."""

    state: MusicRecognitionState
    note: str | None = None

    model_config = ConfigDict(extra="forbid")


class MusicPrepareRequest(BaseModel):
    input: MusicMediaInput

    model_config = ConfigDict(extra="forbid")


class MusicResolveRequest(BaseModel):
    input: MusicMediaInput

    model_config = ConfigDict(extra="forbid")


class MusicPrepareResponse(BaseModel):
    input: MusicMediaInput
    base: MusicMetaBase
    assessment: MusicRecognitionAssessment

    model_config = ConfigDict(extra="forbid")


class MusicResolveResponse(BaseModel):
    base: MusicMetaBase
    assessment: MusicRecognitionAssessment
    media: MusicMediaInfo

    model_config = ConfigDict(extra="forbid")


class MusicResolveDetailRequest(BaseModel):
    input: MusicMediaInput

    model_config = ConfigDict(extra="forbid")


class MusicResolveDetailResponse(BaseModel):
    base: MusicMetaBase
    assessment: MusicRecognitionAssessment
    media: MusicMediaInfo
    detail: MetadataDetail

    model_config = ConfigDict(extra="forbid")
